File: android/app/src/main/java/com/nextpage/data/remote/sync/DriveColdBackupService.kt
package com.nextpage.data.remote.sync

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.nextpage.data.local.dao.BookDao
import com.nextpage.data.local.dao.BookmarkDao
import com.nextpage.data.local.dao.HighlightDao
import com.nextpage.data.local.dao.ReadingProgressDao
import com.nextpage.data.local.dao.ReadingSessionDao
import com.nextpage.data.remote.supabase.BookmarkRow
import com.nextpage.data.remote.supabase.HighlightRow
import com.nextpage.data.remote.supabase.ReadingProgressRow
import com.nextpage.data.remote.supabase.ReadingSessionRow
import com.nextpage.data.remote.supabase.SupabaseBookCatalogDataSource
import com.nextpage.data.remote.supabase.SupabaseProgressDataSource
import com.nextpage.data.remote.supabase.UserBookRow
import com.nextpage.data.session.SessionManager
import kotlinx.coroutines.flow.first
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * DriveColdBackupService — cold backup on demand only (PR3).
 *
 * Reuses [StorageSyncRemoteDataSource] primitives (parents only on create,
 * PATCH without parents keeps 403 parents fix). Export/import are Settings-only;
 * hot save/open never touches Drive.
 *
 * Cold backup file lives as `books/{userId}/nextpage_cold_backup.json`
 * (physical `nextpage_cold_backup.json` inside `NextPage/Books`). JSON+bin shape
 * mirrors desktop parity; bins are book files already in Drive — JSON carries
 * metadata for FK-order restore.
 *
 * Import is FK-ordered `books→progress→highlights→bookmarks→sessions`
 * chunk 100 idempotent (onConflict). Backfill reuses same path.
 */
class DriveColdBackupService(
    private val remoteDataSource: StorageSyncRemoteDataSource,
    private val bookDao: BookDao,
    private val readingProgressDao: ReadingProgressDao,
    private val highlightDao: HighlightDao,
    private val bookmarkDao: BookmarkDao,
    private val readingSessionDao: ReadingSessionDao,
    private val bookCatalogDataSource: SupabaseBookCatalogDataSource = SupabaseBookCatalogDataSource(),
    private val progressDataSource: SupabaseProgressDataSource = SupabaseProgressDataSource(),
    private val sessionManager: SessionManager,
    private val gson: Gson = Gson(),
) {
    companion object {
        const val COLD_BACKUP_FILE = "nextpage_cold_backup.json"
        const val CHUNK_SIZE = 100
        fun coldBackupPath(userId: String) = "books/$userId/$COLD_BACKUP_FILE"
        private fun nowIso(): String = SimpleDateFormat(
            "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US
        ).apply { timeZone = TimeZone.getTimeZone("UTC") }.format(Date())
        private fun millisToIso(millis: Long): String = SimpleDateFormat(
            "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US
        ).apply { timeZone = TimeZone.getTimeZone("UTC") }.format(Date(millis))
    }

    data class ColdBackupJson(
        val version: Int = 1,
        @SerializedName("exported_at") val exportedAt: Long,
        val books: List<UserBookRow> = emptyList(),
        val progress: List<ReadingProgressRow> = emptyList(),
        val highlights: List<HighlightRow> = emptyList(),
        val bookmarks: List<BookmarkRow> = emptyList(),
        val sessions: List<ReadingSessionRow> = emptyList(),
    )

    private suspend fun hasLiveSession(): Boolean =
        sessionManager.getCurrentSession().getOrNull() != null

    /**
     * Export local Room state to Drive as single JSON.
     * Gated only by Drive auth (remoteDataSource will throw on failure);
     * callers (Settings) surface the error. Never called from hot save.
     */
    suspend fun exportColdBackup(userId: String): Result<Unit> = runCatching {
        val books = bookDao.observeAllBooks().first()
        val highlights = highlightDao.observeAllHighlights().first()
        val bookmarks = bookmarkDao.observeAllBookmarks().first()
        val sessions = readingSessionDao.getAll()
        // Progress: one row per book via DAO list or per-book lookup
        val progressEntities = readingProgressDao.getAll()

        val bookRows = books.map { e ->
            UserBookRow(
                id = e.id,
                userId = userId,
                title = e.title,
                author = e.author,
                format = e.format,
                contentHash = e.contentHash,
                filePath = null,
                coverUrl = null,
                description = e.description,
                totalPages = e.totalPages,
                importedAt = millisToIso(e.updatedAtEpochMillis),
                updatedAt = millisToIso(e.updatedAtEpochMillis),
                lifecycle = e.remoteLifecycle.ifEmpty { "available" },
                catalogVersion = maxOf(1L, e.remoteCatalogVersion),
                remoteProvider = e.remoteProvider,
                remoteFileId = e.remoteFileId,
                remotePath = e.remotePath,
            )
        }

        val progressRows = progressEntities.map { e ->
            ReadingProgressRow(
                userId = userId,
                bookId = e.bookId,
                cfiLocation = e.cfiLocation,
                percentage = e.percentage.toDouble(),
                locatorJson = e.locatorJson,
                updatedAt = millisToIso(e.updatedAtEpochMillis),
                version = 1,
            )
        }

        val highlightRows = highlights.map { e ->
            HighlightRow(
                userId = userId,
                bookId = e.bookId,
                cfiRange = e.cfiRange,
                textContent = e.textContent,
                note = e.note,
                color = e.color,
                type = e.type,
                locatorJson = e.locatorJson,
                deletedAt = e.deletedAtEpochMillis?.let { millisToIso(it) },
                updatedAt = millisToIso(e.updatedAtEpochMillis),
                id = e.id,
            )
        }

        val bookmarkRows = bookmarks.map { e ->
            BookmarkRow(
                userId = userId,
                bookId = e.bookId,
                cfiLocation = e.cfiLocation,
                titleSnippet = e.titleOrSnippet,
                locatorJson = e.locatorJson,
                deletedAt = e.deletedAtEpochMillis?.let { millisToIso(it) },
                updatedAt = millisToIso(e.updatedAtEpochMillis),
                id = e.id,
            )
        }

        val sessionRows = sessions.map { e ->
            ReadingSessionRow(
                id = e.id,
                userId = e.userId.ifEmpty { userId },
                bookId = e.bookId,
                startedAt = millisToIso(e.startTimeEpochMillis),
                durationMinutes = e.durationMinutes,
                date = millisToIso(e.date),
                device = "android",
                updatedAt = millisToIso(if (e.updatedAtEpochMillis == 0L) System.currentTimeMillis() else e.updatedAtEpochMillis),
                startPercentage = null,
                endPercentage = null,
            )
        }

        val backup = ColdBackupJson(
            exportedAt = System.currentTimeMillis(),
            books = bookRows,
            progress = progressRows,
            highlights = highlightRows,
            bookmarks = bookmarkRows,
            sessions = sessionRows,
        )
        val jsonBytes = gson.toJson(backup).toByteArray(Charsets.UTF_8)
        // Reuses GoogleDriveStorageRemoteDataSource upload which sends parents only on create,
        // PATCH without parents (403 fix) — drive.file scope.
        remoteDataSource.upload(coldBackupPath(userId), jsonBytes)
    }

    /**
     * Import cold backup JSON from Drive, FK-order chunk 100 idempotent.
     * Gated by hasLiveSession — no request fires without live session.
     */
    suspend fun importColdBackup(userId: String): Result<ImportResult> {
        return runCatching {
            if (!hasLiveSession()) return@runCatching ImportResult(0, 0, 0, 0, 0)
            val bytes = remoteDataSource.download(coldBackupPath(userId))
            val json = String(bytes, Charsets.UTF_8)
            val backup = gson.fromJson(json, ColdBackupJson::class.java)
            importInFkOrder(backup, userId)
        }
    }

    /**
     * One-shot Drive→Supabase backfill, FK-order chunk 100 idempotent, gated on first login.
     * Reuses Drive catalog read: delegates to importColdBackup when cold file exists,
     * otherwise falls back to legacy BookStateJson files listing (no FK errors).
     */
    suspend fun backfillFromDrive(userId: String): Result<ImportResult> {
        return runCatching {
            if (!hasLiveSession()) return@runCatching ImportResult(0, 0, 0, 0, 0)
            // Prefer cold backup; fallback to legacy per-book state.json scan for migrated users
            val coldResult = importColdBackup(userId).getOrNull()
            if (coldResult != null && coldResult.totalImported > 0) return@runCatching coldResult
            // Legacy fallback: list Drive state files prefix and aggregate BookStateJson
            val legacy = collectLegacyDriveState(userId)
            if (legacy != null) importInFkOrder(legacy, userId) else ImportResult(0, 0, 0, 0, 0)
        }
    }

    private suspend fun collectLegacyDriveState(userId: String): ColdBackupJson? {
        return try {
        val prefix = "books/$userId/"
        val paths = remoteDataSource.list(prefix)
        val statePaths = paths.filter { it.endsWith("/state.json") || it.endsWith("_state.json") }
        if (statePaths.isEmpty()) return null
        val gsonLocal = gson
        val bookRows = mutableListOf<UserBookRow>()
        val progressRows = mutableListOf<ReadingProgressRow>()
        val highlightRows = mutableListOf<HighlightRow>()
        val bookmarkRows = mutableListOf<BookmarkRow>()
        // BookStateJson parsing is best-effort; delegate to GoogleDriveJsonStateSync schema
        for (p in statePaths) {
            try {
                val bytes = remoteDataSource.download(p)
                val json = String(bytes, Charsets.UTF_8)
                val state = gsonLocal.fromJson(json, com.nextpage.data.remote.sync.BookStateJson::class.java)
                state.progress?.let { pr ->
                    progressRows.add(
                        ReadingProgressRow(
                            userId = userId, bookId = pr.bookId, cfiLocation = pr.cfiLocation,
                            percentage = pr.percentage.toDouble(), updatedAt = millisToIso(pr.updatedAtEpochMillis), version = 1
                        )
                    )
                }
                state.highlights.forEach { h ->
                    highlightRows.add(
                        HighlightRow(
                            userId = userId, bookId = h.bookId, cfiRange = h.cfiRange,
                            textContent = h.textContent, note = h.note, color = h.color,
                            updatedAt = millisToIso(h.updatedAtEpochMillis),
                            deletedAt = h.deletedAtEpochMillis?.let { millisToIso(it) }, id = h.id
                        )
                    )
                }
                state.bookmarks.forEach { b ->
                    bookmarkRows.add(
                        BookmarkRow(
                            userId = userId, bookId = b.bookId, cfiLocation = b.cfiLocation,
                            titleSnippet = b.titleOrSnippet,
                            updatedAt = millisToIso(b.updatedAtEpochMillis),
                            deletedAt = b.deletedAtEpochMillis?.let { millisToIso(it) }, id = b.id
                        )
                    )
                }
                // Derive stub book rows from state file names when catalog missing
                val bookId = p.substringAfterLast('/').removeSuffix("_state.json").removeSuffix("/state.json")
                if (bookId.isNotBlank() && bookRows.none { it.id == bookId }) {
                    bookRows.add(
                        UserBookRow(
                            id = bookId, userId = userId, title = bookId, format = "epub",
                            importedAt = nowIso(), updatedAt = nowIso()
                        )
                    )
                }
            } catch (_: Exception) { /* skip corrupt file */ }
        }
        if (progressRows.isEmpty() && highlightRows.isEmpty() && bookmarkRows.isEmpty()) null
        else ColdBackupJson(exportedAt = System.currentTimeMillis(), books = bookRows, progress = progressRows, highlights = highlightRows, bookmarks = bookmarkRows)
    } catch (_: Exception) { null }
    }

    private suspend fun importInFkOrder(backup: ColdBackupJson, userId: String): ImportResult {
        var b = 0; var p = 0; var h = 0; var bm = 0; var s = 0
        // 1. books first (FK parent)
        for (chunk in backup.books.chunked(CHUNK_SIZE)) {
            for (row in chunk) {
                try { bookCatalogDataSource.upsertBook(row.copy(userId = userId)); b++ } catch (_: Exception) {}
            }
        }
        // 2. progress (depends on books)
        for (chunk in backup.progress.chunked(CHUNK_SIZE)) {
            for (row in chunk) {
                try { progressDataSource.upsertProgress(row.copy(userId = userId)); p++ } catch (_: Exception) {}
            }
        }
        // 3. highlights
        for (chunk in backup.highlights.chunked(CHUNK_SIZE)) {
            for (row in chunk) {
                try { progressDataSource.upsertHighlight(row.copy(userId = userId)); h++ } catch (_: Exception) {}
            }
        }
        // 4. bookmarks (unique cfi_location)
        for (chunk in backup.bookmarks.chunked(CHUNK_SIZE)) {
            for (row in chunk) {
                try { progressDataSource.upsertBookmark(row.copy(userId = userId)); bm++ } catch (_: Exception) {}
            }
        }
        // 5. sessions last
        for (chunk in backup.sessions.chunked(CHUNK_SIZE)) {
            for (row in chunk) {
                try { progressDataSource.upsertReadingSession(row.copy(userId = userId)); s++ } catch (_: Exception) {}
            }
        }
        return ImportResult(b, p, h, bm, s)
    }

    data class ImportResult(
        val books: Int, val progress: Int, val highlights: Int, val bookmarks: Int, val sessions: Int
    ) {
        val totalImported: Int get() = books + progress + highlights + bookmarks + sessions
    }
}

File: desktop/src/lib/shared/services/DriveColdBackupService.ts
/**
 * DriveColdBackupService — cold backup on demand only (PR3).
 * Reuses GDriveProvider primitives (parents only on create, PATCH without parents
 * keeps 403 parents fix, drive.file scope). Export/import are Settings-only;
 * hot save/open never touches Drive.
 *
 * Cold backup file: `books/{userId}/nextpage_cold_backup.json`
 * (physical `nextpage_cold_backup.json` inside `NextPage/Books`). JSON carries
 * metadata for FK-order restore; bins are Drive book files already in Books folder.
 *
 * Import is FK-ordered `books→progress→highlights→bookmarks→sessions`
 * chunk 100 idempotent (onConflict). Backfill reuses same path.
 */
import { GDriveProvider } from './storage/GDriveProvider';
import { hasLiveSession } from '$lib/services/supabase';
import { SupabaseProgressSync } from '../sync/SupabaseProgressSync';
import { SupabaseBookCatalogSync } from '../sync/SupabaseBookCatalogSync';
import type { RemoteReadingSessionRow } from '$lib/shared/types';
import type { LibraryPort } from '$lib/shared/ports/LibraryPort';
import type { ViewerPort } from '$lib/shared/ports/ViewerPort';
import { TauriLibraryAdapter } from '$lib/shared/ports/adapters/tauri/TauriLibraryAdapter';
import { TauriViewerAdapter } from '$lib/shared/ports/adapters/tauri/TauriViewerAdapter';

export interface ColdBackupJson {
  version: number;
  exportedAt: number;
  books: Array<{
    id: string;
    userId: string;
    title: string;
    author: string | null;
    format: string;
    contentHash?: string | null;
    importedAt: string;
    updatedAt: string;
  }>;
  progress: Array<{
    userId: string;
    bookId: string;
    cfiLocation: string;
    percentage: number;
    locatorJson?: string | null;
    updatedAt: string;
  }>;
  highlights: Array<{
    id: string;
    userId: string;
    bookId: string;
    cfiRange: string;
    textContent: string;
    note?: string | null;
    color: string;
    updatedAt: string;
    deletedAt?: string | null;
    locatorJson?: string | null;
  }>;
  bookmarks: Array<{
    id: string;
    userId: string;
    bookId: string;
    cfiLocation: string;
    titleSnippet?: string | null;
    locatorJson?: string | null;
    updatedAt: string;
    deletedAt?: string | null;
  }>;
  sessions: RemoteReadingSessionRow[];
}

export interface ImportResult {
  books: number;
  progress: number;
  highlights: number;
  bookmarks: number;
  sessions: number;
  totalImported: number;
}

const CHUNK_SIZE = 100;
const COLD_BACKUP_FILE = 'nextpage_cold_backup.json';

function chunk<T>(arr: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < arr.length; i += size) out.push(arr.slice(i, i + size));
  return out;
}

export class DriveColdBackupService {
  private static libraryPort: LibraryPort = new TauriLibraryAdapter();
  private static viewerPort: ViewerPort = new TauriViewerAdapter();
  static setLibraryPort(port: LibraryPort): void { this.libraryPort = port; }
  static setViewerPort(port: ViewerPort): void { this.viewerPort = port; }
  private static gdrive = new GDriveProvider();

  /**
   * Export local state to Drive as single JSON.
   * Gathers via tauri (books, progress, highlights, bookmarks) and Supabase sessions (if live).
   * Reuses GDriveProvider.upload which sends parents only on create (403 fix).
   */
  static async exportColdBackup(userId: string): Promise<void> {
    const [libraryBooks, sourceBooks] = await Promise.all([
      this.libraryPort.listLibraryBooks().catch(() => []),
      this.libraryPort.listBooks().catch(() => []),
    ]);

    // Books: map library DTOs to catalog rows
    const books: ColdBackupJson['books'] = libraryBooks.map((b: import('$lib/shared/types').LibraryBookDto) => ({
      id: b.id,
      userId,
      title: b.title,
      author: b.author ?? null,
      format: b.format,
      contentHash: null,
      importedAt: b.createdAt ?? new Date().toISOString(),
      updatedAt: b.updatedAt ?? new Date().toISOString(),
    }));

    // Progress: one per book
    const progress: ColdBackupJson['progress'] = [];
    for (const b of sourceBooks) {
      try {
        const p = await this.viewerPort.getProgress(b.id);
        if (p) {
          progress.push({
            userId,
            bookId: p.bookId,
            cfiLocation: p.cfiLocation,
            percentage: p.percentage,
            updatedAt: p.updatedAt,
          });
        }
      } catch {
        // skip
      }
    }

    // Highlights + bookmarks per book
    const highlights: ColdBackupJson['highlights'] = [];
    const bookmarks: ColdBackupJson['bookmarks'] = [];
    for (const b of sourceBooks) {
      try {
        const hs = await this.viewerPort.listHighlights(b.id);
        for (const h of hs) {
          highlights.push({
            id: h.id,
            userId,
            bookId: h.bookId,
            cfiRange: h.cfi ?? '',
            textContent: h.text,
            note: h.note ?? null,
            color: h.color,
            updatedAt: h.createdAt ?? new Date().toISOString(),
            deletedAt: null,
          });
        }
      } catch {}
      try {
        const bms = await this.viewerPort.listBookmarks(b.id);
        for (const bm of bms) {
          bookmarks.push({
            id: bm.id,
            userId,
            bookId: bm.bookId,
            cfiLocation: '',
            titleSnippet: bm.title ?? null,
            updatedAt: bm.createdAt ?? new Date().toISOString(),
            deletedAt: null,
          });
        }
      } catch {}
    }

    // Sessions: fetch from Supabase when live, else empty (local sessions already on disk)
    let sessions: RemoteReadingSessionRow[] = [];
    if (hasLiveSession()) {
      try {
        const sync = new SupabaseProgressSync(userId);
        sessions = await sync.fetchReadingSessions();
      } catch {}
    }

    const backup: ColdBackupJson = {
      version: 1,
      exportedAt: Date.now(),
      books,
      progress,
      highlights,
      bookmarks,
      sessions,
    };

    const jsonBytes = new TextEncoder().encode(JSON.stringify(backup));
    // Reuses GDriveProvider.upload: POST with parents on create, PATCH without parents on update (403 fix)
    await this.gdrive.upload(COLD_BACKUP_FILE, jsonBytes as unknown as Uint8Array, COLD_BACKUP_FILE);
  }

  /**
   * Import cold backup JSON from Drive, FK-order chunk 100 idempotent.
   * Gated by hasLiveSession — no request fires without live session.
   */
  static async importColdBackup(userId: string): Promise<ImportResult> {
    if (!hasLiveSession()) return { books: 0, progress: 0, highlights: 0, bookmarks: 0, sessions: 0, totalImported: 0 };
    const bytes: Uint8Array = await this.gdrive.download(COLD_BACKUP_FILE);
    const json = new TextDecoder().decode(bytes);
    const backup = JSON.parse(json) as ColdBackupJson;
    return this.importInFkOrder(backup, userId);
  }

  /**
   * One-shot Drive→Supabase backfill, FK-order chunk 100 idempotent, gated on first login.
   * Reuses Drive catalog read: delegates to importColdBackup when cold file exists,
   * otherwise no-op (legacy per-book state.json path already covered by import).
   */
  static async backfillFromDrive(userId: string): Promise<ImportResult> {
    if (!hasLiveSession()) return { books: 0, progress: 0, highlights: 0, bookmarks: 0, sessions: 0, totalImported: 0 };
    try {
      const res = await this.importColdBackup(userId);
      if (res.totalImported > 0) return res;
    } catch {}
    // Legacy fallback: list Drive state files via GDriveProvider.list and aggregate
    try {
      const files = await this.gdrive.list('');
      const stateFiles = files.filter((f) => f.endsWith('_state.json'));
      if (stateFiles.length === 0) return { books: 0, progress: 0, highlights: 0, bookmarks: 0, sessions: 0, totalImported: 0 };
      // Best-effort: each state.json is BookStateJson with progress/highlights/bookmarks
      const backup: ColdBackupJson = { version: 1, exportedAt: Date.now(), books: [], progress: [], highlights: [], bookmarks: [], sessions: [] };
      for (const f of stateFiles) {
        try {
          const b = await this.gdrive.download(f);
          const j = JSON.parse(new TextDecoder().decode(b)) as { progress: { book_id: string; cfi_location: string; percentage: number; updated_at: number } | null; highlights: Array<{ id: string; book_id: string; cfi_range: string; text_content: string; note: string | null; color: string; updated_at: number }>; bookmarks: Array<{ id: string; book_id: string; cfi_location: string; title_or_snippet: string; updated_at: number }> };
          if (j.progress) {
            backup.progress.push({
              userId,
              bookId: j.progress.book_id,
              cfiLocation: j.progress.cfi_location,
              percentage: j.progress.percentage,
              updatedAt: new Date(j.progress.updated_at).toISOString(),
            });
          }
          for (const h of j.highlights ?? []) {
            backup.highlights.push({
              id: h.id,
              userId,
              bookId: h.book_id,
              cfiRange: h.cfi_range,
              textContent: h.text_content,
              note: h.note,
              color: h.color,
              updatedAt: new Date(h.updated_at).toISOString(),
            });
          }
          for (const bm of j.bookmarks ?? []) {
            backup.bookmarks.push({
              id: bm.id,
              userId,
              bookId: bm.book_id,
              cfiLocation: bm.cfi_location,
              titleSnippet: bm.title_or_snippet,
              updatedAt: new Date(bm.updated_at).toISOString(),
            });
          }
        } catch {}
      }
      if (backup.progress.length === 0 && backup.highlights.length === 0 && backup.bookmarks.length === 0) {
        return { books: 0, progress: 0, highlights: 0, bookmarks: 0, sessions: 0, totalImported: 0 };
      }
      return this.importInFkOrder(backup, userId);
    } catch {
      return { books: 0, progress: 0, highlights: 0, bookmarks: 0, sessions: 0, totalImported: 0 };
    }
  }

  private static async importInFkOrder(backup: ColdBackupJson, userId: string): Promise<ImportResult> {
    let books = 0;
    let progress = 0;
    let highlights = 0;
    let bookmarks = 0;
    let sessions = 0;

    const bookSync = new SupabaseBookCatalogSync(userId);
    const progressSync = new SupabaseProgressSync(userId);

    // 1. books first (FK parent)
    for (const c of chunk(backup.books ?? [], CHUNK_SIZE)) {
      for (const row of c) {
        try {
          await bookSync.upsertBook({
            id: row.id,
            userId,
            title: row.title,
            author: row.author,
            format: row.format,
            contentHash: row.contentHash ?? null,
            filePath: null,
            coverUrl: null,
            description: null,
            totalPages: null,
            sourceDevice: 'desktop',
            importedAt: row.importedAt,
            updatedAt: row.updatedAt,
          });
          books++;
        } catch {}
      }
    }
    // 2. progress
    for (const c of chunk(backup.progress ?? [], CHUNK_SIZE)) {
      for (const row of c) {
        try {
          await progressSync.upsertProgress({
            userId,
            bookId: row.bookId,
            cfiLocation: row.cfiLocation,
            percentage: row.percentage,
            updatedAt: row.updatedAt,
          });
          progress++;
        } catch {}
      }
    }
    // 3. highlights
    for (const c of chunk(backup.highlights ?? [], CHUNK_SIZE)) {
      for (const row of c) {
        try {
          await progressSync.upsertHighlight({
            userId,
            bookId: row.bookId,
            cfiRange: row.cfiRange,
            textContent: row.textContent,
            note: row.note ?? null,
            color: row.color,
            updatedAt: row.updatedAt,
            deletedAt: row.deletedAt ?? null,
          } as never);
          highlights++;
        } catch {}
      }
    }
    // 4. bookmarks
    for (const c of chunk(backup.bookmarks ?? [], CHUNK_SIZE)) {
      for (const row of c) {
        try {
          await progressSync.upsertBookmark({
            userId,
            bookId: row.bookId,
            cfiLocation: row.cfiLocation,
            titleSnippet: row.titleSnippet ?? null,
            updatedAt: row.updatedAt,
            deletedAt: row.deletedAt ?? null,
          } as never);
          bookmarks++;
        } catch {}
      }
    }
    // 5. sessions last (chunk 100, idempotent onConflict id)
    for (const c of chunk(backup.sessions ?? [], CHUNK_SIZE)) {
      for (const row of c) {
        try {
          await progressSync.upsertReadingSession({
            id: row.id,
            userId,
            bookId: row.bookId,
            startedAt: row.startedAt,
            durationMinutes: row.durationMinutes,
            date: row.date,
            device: 'desktop',
            updatedAt: new Date(row.updatedAtEpochMillis).toISOString(),
            startPercentage: row.startPercentage ?? null,
            endPercentage: row.endPercentage ?? null,
          });
          sessions++;
        } catch {}
      }
    }

    const totalImported = books + progress + highlights + bookmarks + sessions;
    return { books, progress, highlights, bookmarks, sessions, totalImported };
  }
}
